### ISSUE TITLE
Nested object Arrays

### ISSUE THREAD

**Issue by mrchief** [mholt/PapaParse#266]
Is it possible to support arrays of nested objects?

E.g., given input like this:

```
fieldA.title, fieldA.children[0].name, fieldA.children[0].id,fieldA.children[1].name, fieldA.children[1].employee[].name,fieldA.children[1].employee[].name, fieldA.address[],fieldA.address[], description
Food Factory, Oscar, 0023, Tikka, Tim, Joe, 3 Lame Road, Grantstown, A fresh new food factory
Kindom Garden, Ceil, 54, Pillow, Amst, Tom, 24 Shaker Street, HelloTown, Awesome castle
```

turn into 

```json
[{
    "fieldA": {
        "title": "Food Factory",
        "children": [{
            "name": "Oscar",
            "id": "0023"
        }, {
            "name": "Tikka",
            "employee": [{
                "name": "Tim"
            }, {
                "name": "Joe"
            }]
        }],
        "address": ["3 Lame Road", "Grantstown"]
    },
    "description": "A fresh new food factory"
}, {
    "fieldA": {
        "title": "Kindom Garden",
        "children": [{
            "name": "Ceil",
            "id": "54"
        }, {
            "name": "Pillow",
            "employee": [{
                "name": "Amst"
            }, {
                "name": "Tom"
            }]
        }],
        "address": ["24 Shaker Street", "HelloTown"]
    },
    "description": "Awesome castle"
}]
```

Similar to what this does: https://github.com/Keyang/node-csvtojson#empowered-json-parser

**Comment by mholt** [comment]
When you ask, "is it possible" are you _really_ asking if it's possible? Of course the answer is yes. JavaScript is Turing-complete and we can do anything. But if you're asking if it's within the scope of this project to support arrays of nested objects, then the answer is no.

(Sorry for the snark; I always get questions starting with "is it possible" and I just couldn't resist.)

This is similar to https://github.com/mholt/PapaParse/issues/134. Basically, CSV is by nature a flat file format and Papa Parse is a pure CSV parser; it doesn't make any attempt to interpret the data. Also there are soooo many ways to flatten complex data structures and trying to support it all would be impractical. But you should be able to take its output and transform it further if needed.

Thanks for asking! Maybe I will add this to FAQ.

**Comment by mrchief** [comment]
I saw 134 before posting this. The difference in my case is that I'm not asking for any specific interpretation which requires understanding of underlying data like 134. 

You may call evaluating `parent["child[0].xxx"]` as `parent.child[0].name` as interpretation but then, the difference is that the data is self describing and all information (index, prop name) is present in the key itself. Its a simple convention which is independent of underlying data's structure. 

I think the dots are leading you to believe that my ask is similar to 134. In that case, I'm open to change the convention, e.g.;

```
childName, childAge, childName, childAge
John, 12, Stacy, 5
```

resulting in

```JSON
{
  childName: ["John", "Stacy"],
  childAge: [12, 5]
}
``` 

However, this would still require "some" manual labor on the caller's part to turn it into (splitting at the camel casing humps) 

```json
{
  child: [{Name: "John", Age: 12}, {Name: "Stacy", Age: 5} ]
}
```
which is better than nothing since they have somewhat structured data to begin with.

But I also see how you could shoot this down as, if they can do this much, they surely can do a bit more and thus leave PapaParse alone. Fair enough I would say.

So in summary, yeah I understand where you're coming from and why you resist this. 

It'd be nice though if we could get hooks early into the parsing process like `afterHeaderParse` (to create initial object structure based on expected outcome) and `beforeRowPush` (to push data into the expected structure instead of the flat array) so that we can manage these manual transformations more easily.

**Comment by mholt** [comment]
Those hooks I _will_ consider; evaluating dot and array notation out of strings, probably not. :wink:

**Comment by allenfuller** [comment]
+1 to the hooks. I'm using PapaParse within Meteor and my schema has nested arrays. This is such a great, fast parser, it would be great to be able to fully use it in that context.
